include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -0,0 +1,522 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <ginkgo/core/base/dense_cache.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/base.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+template <typename ValueType, typename IndexType>
+class Csr;
+
+
+}
+
+
+namespace detail {
+
+
+template <template <typename, typename> class MatrixType,
+          typename... CreateArgs>
+struct MatrixTypeBuilderFromValueAndIndex {
+    template <typename ValueType, typename IndexType, std::size_t... I>
+    auto create_impl(std::shared_ptr<const Executor> exec,
+                     std::index_sequence<I...>)
+    {
+        return MatrixType<ValueType, IndexType>::create(
+            exec, std::get<I>(create_args)...);
+    }
+
+
+    template <typename ValueType, typename IndexType>
+    auto create(std::shared_ptr<const Executor> exec)
+    {
+        // with c++17 we could use std::apply
+        static constexpr auto size = sizeof...(CreateArgs);
+        return create_impl<ValueType, IndexType>(
+            std::move(exec), std::make_index_sequence<size>{});
+    }
+
+    std::tuple<CreateArgs...> create_args;
+};
+
+
+template <typename T, typename = void>
+struct is_matrix_builder : std::false_type {};
+
+template <typename T>
+struct is_matrix_builder<
+    T, xstd::void_t<decltype(std::declval<T>().template create<double, int>(
+           std::declval<std::shared_ptr<const Executor>>()))>>
+    : std::true_type {};
+
+
+}  // namespace detail
```

Comment:
Since the latter one does not repeat the value and index type, I think it is worth it. Also, I will remove the `is_matrix_builder` stuff, since I didn't end up using it.

Reply:
The main issue I see with it is that it introduces another unfamiliar idiom and an associated free function that is only used in this one constructor. The difference in amount and complexity of code you need to write is marginal. But as I said above, I don't want to hold the PR up over this, especially since the effort was already spent :)